envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotUpdaterTest.kt

```diff
@@ -262,6 +263,93 @@ class SnapshotUpdaterTest {
         results[1].xdsSnapshot!!.hasHttp2Cluster("service")
     }
 
+    @Test
+    fun `should not change EDS when remote doesn't have state of service`() {
+        // given
+        val cache = MockCache()
+        val allServicesGroup = AllServicesGroup(communicationMode = ADS)
+        val groups = listOf(allServicesGroup)
+        val updater = snapshotUpdater(
+            cache = cache,
+            properties = SnapshotProperties().apply {
+                stateSampleDuration = Duration.ZERO
+            },
+            groups = groups
+        )
+
+        val clusterLocal = ClusterState(
+            ServicesState(
+                serviceNameToInstances = mapOf(
+                    "service" to ServiceInstances("service", setOf(ServiceInstance(
+                        id = "id",
+                        tags = setOf("envoy"),
+                        address = "127.0.0.3",
+                        port = 4444
+                    ))),
+                    "service2" to ServiceInstances("service2", setOf())
+                )
+            ),
+            Locality.LOCAL, "cluster"
+        ).toMultiClusterState()
+
+        val clusterRemote1 = ClusterState(
+            ServicesState(
+                serviceNameToInstances = mapOf(
+                    "service" to ServiceInstances("service", setOf()),
+                    "service2" to ServiceInstances("service2", setOf())
+                )
+            ),
+            Locality.REMOTE, "cluster2"
+        ).toMultiClusterState()
+
+        val clusterRemote2 = ClusterState(
+            ServicesState(
+                serviceNameToInstances = mapOf(
+                    "service" to ServiceInstances("service", setOf())
+                )
+            ),
+            Locality.REMOTE, "cluster2"
+        ).toMultiClusterState()
+
+        // when
+        val results = updater
```

Comment:
You could use [Reactor testing tools](https://projectreactor.io/docs/core/release/reference/#testing) instead of this dance of delaying, blocking, etc.

Reply:
We might refactor tests to use [Reactor testing tools](https://projectreactor.io/docs/core/release/reference/#testing) in future.